src/ice/agent.rs

```diff
@@ -761,63 +761,82 @@ impl IceAgent {
-                    local,
-                    local_creds.ufrag
-                );
-                return false;
+        let do_integrity_check = |is_request: bool| -> bool {
+            let (_, password) = self.stun_credentials(is_request);
+            let integrity_passed = message.check_integrity(&password);
+
+            // The integrity is always the last thing we check
+            if integrity_passed {
+                trace!("Message accepted");
+            } else {
+                trace!("Message rejected, integrity check failed");
             }
+            integrity_passed
+        };
 
-            if let Some(remote_creds) = &self.remote_credentials {
-                if remote != remote_creds.ufrag {
+        let method = message.method();
+        let class = message.class();
+        match (method, class) {
+            (StunMethod::Binding, StunClass::Request | StunClass::Indication) => {
+                // The username for the credential is formed by concatenating the
+                // username fragment provided by the peer with the username fragment of
+                // the ICE agent sending the request, separated by a colon (":").
+                // The existence of this username is checked in the STUN parser.
+                let (local, remote) = message.split_username().unwrap();
+
+                let local_creds = self.local_credentials();
+                if local != local_creds.ufrag {
                     trace!(
-                        "Message rejected, remote user mismatch: {} != {}",
-                        remote,
-                        remote_creds.ufrag
+                        "Message rejected, local user mismatch: {} != {}",
+                        local,
+                        local_creds.ufrag
                     );
                     return false;
                 }
-            }
-        }
 
-        if message.is_successful_binding_response() {
-            let belongs_to_a_candidate_pair = self
-                .candidate_pairs
-                .iter()
-                .any(|pair| pair.has_binding_attempt(message.trans_id()));
+                if let Some(remote_creds) = &self.remote_credentials {
+                    if remote != remote_creds.ufrag {
+                        trace!(
+                            "Message rejected, remote user mismatch: {} != {}",
+                            remote,
+                            remote_creds.ufrag
+                        );
+                        return false;
+                    }
+                }
 
-            if !belongs_to_a_candidate_pair {
-                trace!("Message rejected, transaction ID does not belong to any of our candidate pairs");
-                return false;
+                do_integrity_check(true)
             }
-        }
+            (StunMethod::Binding, StunClass::Success | StunClass::Failure) => {
+                let belongs_to_a_candidate_pair = self
```

Comment:
Maybe we should check nominated pair here first, to make a less compute intensive happy path?

Reply:
Ah, yes, you're right! We even talked about that in the other PR, I guess it got lost amongst the other discussion happening there. I'll do that.

Reply:
I've looked deeper into the code, and while adding this fast path is possible, it's surprisingly risky and would be better left to a separate PR, if we deem it necessary at all.

The issue is that our nominated pair `IceAgent.nominated_send` is not an index into the pairs array, but rather a `PairId`. Resolving that to an index would require a linear scan of the pairs array anyway, negating any benefits of checking it first. Storing an index instead would be very difficult because we sort pairs by priority and prune failed pairs often, making the bookkeeping very error-prone.

Reply:
Alright! Thanks for checking!